### DISCUSSION TITLE
Script API and Wallclock time

### DISCUSSION THREAD

**Discussion by frosch123** [OpenTTD/OpenTTD#12339] [General]
There are some rumours and myths about the Script API, and how it interacts with the new wallclock time.

Current status:
* ScriptDate returns EconomyTime by default. ScriptTimeMode can temporarily change it to CalendarTime.
* Other API methods return either EconomyTime or CalendarTime, since only one makes sense in each case.
* Most scripts use EconomyTime, CalendarTime is actually rarely needed. (citation needed)

General problems are:
* API methods do not document, which time they use. The documentation should be enhanced.
* GS often show dates and time-spans to the user. While the GS are likely to technically work with EconomyTime, the display is confusing. Nothing we can really fix, the GS must be updated.
* Some scripts assume the absolute year to be in a certain range, e.g. 4 digits, or 1900+.

**Comment by frosch123** [comment]
API methods using calendar dates:
* ScriptBaseStation.GetConstructionDate
* ScriptEngine.GetDesignDate
* ScriptIndustry.GetConstructionDate
* ScriptStoryPage.GetDate
* ScriptStoryPage.SetDate

**Reply by TrueBrain** [reply]
Usage numbers, based on "grep" over all sources.

```
GetConstructionDate
      4 ai
      3 game-script
GetDesignDate
     11 ai
      3 game-script
GetDate
     21 ai-library
     73 ai
     13 game-script-library
     58 game-script
SetDate
     10 game-script
---
Station.GetConstructionDate(
      4 ai
      1 game-script
Engine.GetDesignDate(
      5 ai
      1 game-script
Industry.GetConstructionDate(
StoryPage.GetDate(
StoryPage.SetDate(
      9 game-script
```

`GetDate` is an invalid result, as my simple grep returned more hits than required.

**Reply by frosch123** [reply]
The purpose of these two methods is unclear, and so is their time base:
* ``ScriptBaseStation.GetConstructionDate``
    * This property is reset to current date, whenever the station is modified.
    * Awards GS actually uses it to detect "long service", which assumes a different meaning.
    * Some AIs use it to avoid touching a new station, and wait until service is stable.
    * Both use-cases assume economy-time, but one assumes "first construction", the other "last modification".
* ``ScriptIndustry.GetConstructionDate``
    * Currently unused by any script.
    * There is also ``ScriptIndustry::GetLastProductionYear`` which uses economy-time, and has a meaning for default processing industries.

So, maybe both methods should just be deprecated.

**Comment by frosch123** [comment]
API methods using time-spans from calendar:
* ScriptEngine.GetMaxAge
* ScriptVehicle.GetMaxAge
* ScriptVehicle.GetWagonAge
* ScriptVehicle.GetAge
* ScriptVehicle.GetAgeLeft
* ScriptOrder.OrderCondition.OC_AGE
* ScriptOrder.OrderCondition.OC_REMAINING_LIFETIME

**Reply by TrueBrain** [reply]
Usage numbers, based on "grep" over all sources.

```
GetMaxAge
      1 ai-library
     19 ai
      1 game-script-library
      3 game-script
GetAge
     49 ai
      3 game-script
GetAgeLeft
     25 ai
      2 game-script
OC_AGE
      5 ai
      2 game-script
OC_REMAINING_LIFETIME
      2 ai
      2 game-script
---
Engine.GetMaxAge(
      1 ai-library
      9 ai
      1 game-script-library
Vehicle.GetMaxAge(
     11 ai
      1 game-script
Vehicle.GetAge(
     27 ai
Vehicle.GetAgeLeft(
     12 ai
OC_AGE
      5 ai
      2 game-script
OC_REMAINING_LIFETIME
      2 ai
      2 game-script
```

**Comment by frosch123** [comment]
API methods depending on ScriptTimeMode:
* ScriptDate.GetCurrentDate
* ScriptSubsidy::GetExpireDate: actually broken, should always use economy dates.

**Reply by frosch123** [reply]
@2TallTyler Just checking my understanding: Can you confirm?
* Currently ScriptSubsidy::GetExpireDate adds economy-months to ScriptDate::GetDate.
* If ScriptTimeMode is set to CalendarTime, this will add economy-months to a calendar-date, which makes no sense.
* Instead the method should always return an economy-date.

**Reply by TrueBrain** [reply]
Usage numbers, based on "grep" over all sources.

```
GetCurrentDate
      3 ai-library
     49 ai
      2 game-script-library
     45 game-script
GetExpireDate
      7 ai
      2 game-script
---
Date.GetCurrentDate(
      3 ai-library
     48 ai
      2 game-script-library
     43 game-script
Subsidy.GetExpireDate(
      5 ai
```

**Reply by 2TallTyler** [reply]
Yes, your diagnosis is correct. Everything related to ScriptSubsidy should be in economy time.

**Comment by frosch123** [comment]
API methods using economy dates:
* ScriptIndustry.GetLastProductionYear
* ScriptIndustry.GetCargoLastAcceptedDate
* ScriptClient.GetJoinDate

**Reply by TrueBrain** [reply]
Usage numbers, based on "grep" over all sources.

```
GetLastProductionYear
      1 game-script
GetCargoLastAcceptedDate
      1 game-script
GetJoinDate
      1 game-script
---
Industry.GetLastProductionYear(
Industry.GetCargoLastAcceptedDate(
Client.GetJoinDate(
```

All three the same game-script, named `RemoteGS`.

**Comment by frosch123** [comment]
API suggestion: (revival of #11588)
* Deprecate ScriptTimeMode, remove it in OpenTTD 15.
* Deprecate ScriptDate, remove it in OpenTTD 15.
* Add ScriptCalendarDate and ScriptEconomyDate instead.
    * Script*Date::GetDaysInYear or similar to assist converting economy-days into economy-years, as returned by various API methods.
* Compatiblity scripts continue to provide ScriptDate.

Reasoning:
* Switching TimeMode only affects ScriptDate, so it is not really a global operation.
* Switching TimeMode only for ScriptDate is error-prone.
* Other API methods use fixed timemodes, using separate ScriptCalendarDate and ScriptEconomyDate makes it easier to compare dates with values returnes from other API methods.

**Reply by PeterN** [reply]
To be honest, if ScriptTimeMode is to be removed, I think it should just go instead of leaving it around as deprecated. It's only been in the API for a short amount of time, and not in any proper release.

**Reply by andythenorth** [reply]
Tend to agree.  I've implemented ScriptTimeMode in FIRS GS (to work around Industry construction date), but that's unreleased.  Have any other scripts appeared on Bananas with it?

**Comment by frosch123** [comment]
API methods in context of economy-years:
* ScriptEventEnginePreview.GetRunningCost
* ScriptEngine.GetRunningCost
* ScriptVehicle.GetRunningCost
* ScriptVehicle.GetProfitThisYear
* ScriptVehicle.GetProfitLastYear
* ScriptGroup.GetProfitThisYear
* ScriptGroup.GetProfitLastYear

API methods in context of economy-quarters:
* ScriptCompany.GetQuarterlyIncome
* ScriptCompany.GetQuarterlyExpenses
* ScriptCompany.GetQuarterlyCargoDelivered
* ScriptCompany.GetQuarterlyPerformanceRating
* ScriptCompany.GetQuarterlyCompanyValue

API methods in context of economy-months:
* ScriptInfrastructure.GetMonthlyRailCosts
* ScriptInfrastructure.GetMonthlyRoadCosts
* ScriptInfrastructure.GetMonthlyInfrastructureCosts
* ScriptAirport.GetMonthlyMaintenanceCost
* ScriptIndustry.GetLastMonthProduction
* ScriptIndustry.GetLastMonthTransported
* ScriptIndustry.GetLastMonthTransportedPercentage
* ScriptTown.GetLastMonthProduction
* ScriptTown.GetLastMonthSupplied
* ScriptTown.GetLastMonthTransportedPercentage
* ScriptTown.GetLastMonthReceived
* ScriptTown.GetRoadReworkDuration
* ScriptTown.GetFundBuildingsDuration
* ScriptTown.GetExclusiveRightsDuration
* ScriptTown.SetCargoGoal
* ScriptTown.GetCargoGoal

API methods in context of economy-days:
* ScriptTown.SetGrowthRate
* ScriptTown.GetGrowthRate

**Comment by frosch123** [comment]
API methods in context of calendar-months:
* ScriptCompany.SetAutoRenewMonths
* ScriptCompany.GetAutoRenewMonths

**Comment by MontyMontana** [comment]
Is there any way for a script to get the time mode being used on the game? 

I have searched on the [API documentation](https://docs.openttd.org/gs-api/game_changelog) page and at least is not reflected on changelog. 

It could be interesting (suggestion here) to create a function for gamescripts for letting the script know if timekeeping units are being used, and how minutes per year setting is being used. The only way I currently know for setting this data into the script implementation is via parameters, which means they have to be set and changed it for every game.

**Reply by frosch123** [reply]
@MontyMontana Via ``GSController.GetSetting`` you can read every game setting.
E.g. ``GSController.GetSetting("timekeeping_units")``

**Reply by MontyMontana** [reply]
It's weird, @frosch123. Let me explain.

I have tested with that method, just retrieving game settings and logging them, and it does not work. I have test the one we are talking about, also with "prefix" (economy.timekeeping_units), and other settings too and nothing was returned. I think this function can not do such task. In the [documentation](https://docs.openttd.org/gs-api/classGSController#ac238c3457609bac0ad57484300150456) is written that it can be used only for user-created settings:

> GetSetting()
> 
> Get the value of ** one of your settings you set via info.nut**. 
> 
> Returns
>             the value for the setting, or -1 if the setting is not known. 

However, I did not noticed retrieving a "-1" neither. I will perform more tests, anyway. But I think its not possible to get game settings.

**Reply by stormcone** [reply]
There is a [GSGameSettings](https://docs.openttd.org/gs-api/classGSGameSettings). Maybe this is what you are looking for.

**Reply by glx22** [reply]
Yeah use `GSGameSettings` for game settings, `GSController.GetSetting()` is for the script ones (declared in info.nut).
And always protect `GSGameSettings.GetValue()` with `GSGameSettings.IsValid()`.

**Reply by frosch123** [reply]
Yes, sorry, I linked the wrong method.

**Comment by MontyMontana** [comment]
@stormcone Yes, this one solves the need.
@glx22 That distinction solves our previous confusion, and thanks for the tip with isValid method!
@frosch123 No problem! Any help is always welcome!
Thanks to all three of you for your support!
